### PR TITLE
Transaction Receipt handling

### PR DESCRIPTION
**Pull request by eadventurous**
Signed-off-by: Egor Ivkov <e.o.ivkov@gmail.com>

### Description of the Change
Introduces receipts from leader on getting forwarded transaction. Also handles the case when the leader does not send a receipt.

### CODE REVIEW

File: iroha/src/crypto.rs [hunk 1]
```diff
@@ -57,6 +58,22 @@ pub struct Signature {
...
+            self.public_key.cmp(&other.public_key)
+        }
+    }
+}
+
+impl PartialOrd for Signature {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
```
Comment (humb1t) [inline iroha/src/crypto.rs:75]: Let's keep traits impls under the impl

Comment (eadventurous) [reply iroha/src/crypto.rs]: Sure. Actually this impl is no longer needed so I removed it.

File: iroha/src/sumeragi.rs [hunk 2]
```diff
@@ -23,6 +24,9 @@ use std::{
 /// Amount of time Peer waits for CommitMessage from the proxy tail.
 pub const COMMIT_TIME: Duration = Duration::from_millis(1000);
 
+/// Amount of time Peer waits for TxReceipt from the leader.
+pub const TX_RECEIPT_TIME: Duration = Duration::from_millis(100);
```
Comment (humb1t) [inline iroha/src/sumeragi.rs:28]: We need to provide configuration of such parameters in the same PR they arrive to Iroha.
Please update config.

File: iroha/src/sumeragi.rs [hunk 3]
```diff
@@ -31,59 +35,126 @@ trait Consensus {
 #[derive(Io, Decode, Encode, Debug, Clone)]
 pub enum Message {
     /// Is sent by leader to all validating peers, when a new block is created.
-    Created(SignedBlock),
+    CreatedBlock(SignedBlock),
```
Comment (humb1t) [inline iroha/src/sumeragi.rs:38]: To read it better use `BlockCreated` - otherwise it looks like type of the block.

File: iroha/src/sumeragi.rs [hunk 4]
```diff
@@ -31,59 +35,126 @@ trait Consensus {
...
     /// Is sent by validating peers to proxy tail and observing peers when they have signed this block.
-    Signed(SignedBlock),
+    SignedBlock(SignedBlock),
```
Comment (humb1t) [inline iroha/src/sumeragi.rs:40]: `BlockSigned`

File: iroha/src/sumeragi.rs [hunk 5]
```diff
@@ -31,59 +35,126 @@ trait Consensus {
...
     /// Is sent by proxy tail to validating peers and to leader, when the block is committed.
-    Committed(SignedBlock),
+    CommittedBlock(SignedBlock),
```
Comment (humb1t) [inline iroha/src/sumeragi.rs:42]: `BlockCommited`

File: iroha/src/sumeragi.rs [hunk 6]
```diff
@@ -31,59 +35,126 @@ trait Consensus {
...
     /// Is sent when the node votes to change view due to commit timeout.
     CommitTimeout(CommitTimeout),
+    /// Receipt of receiving tx from peer. Sent by a leader.
+    ReceivedTx(TxReceipt),
```
Comment (humb1t) [inline iroha/src/sumeragi.rs:46]: `TransactionReceived`

File: iroha/src/sumeragi.rs [hunk 7]
```diff
@@ -31,59 +35,126 @@ trait Consensus {
...
+    /// Tx forwarded from client by a peer to a leader.
+    ForwardedTx(ForwardedTx),
```
Comment (humb1t) [inline iroha/src/sumeragi.rs:48]: `TransactionForwarded`

File: iroha/src/sumeragi.rs [hunk 8]
```diff
@@ -31,59 +35,126 @@ trait Consensus {
...
+    /// Message to other peers that this peer did not receive receipt from leader for a forwarded tx.
+    NoTxReceipt(NoTxReceipt),
```
Comment (humb1t) [inline iroha/src/sumeragi.rs:50]: `NoTransactionReceiptReceived`

File: iroha/src/sumeragi.rs [hunk 9]
```diff
@@ -31,59 +35,126 @@ trait Consensus {
...
             &peer.address,
-            Request::new(uri::BLOCKS_URI.to_string(), self.into()),
+            Request::new(uri::CONSENSUS_URI.to_string(), self.into()),
         )
         .await?;
         Ok(())
     }
 }
 
+/// Message structure describing a failed attempt to forward tx to a leader.
```
Comment (humb1t) [inline iroha/src/sumeragi.rs:64]: We have too much messages in the top module, move everything message related into `sumeragi::message` submodule.

Comment (humb1t) [inline iroha/src/sumeragi.rs:64]: `transaction`

File: iroha/src/sumeragi.rs [hunk 10]
```diff
@@ -31,59 +35,126 @@ trait Consensus {
...
     /// This way peers vote for changing the view.
     pub fn sign(&mut self, public_key: &PublicKey, private_key: &PrivateKey) -> Result<(), String> {
-        let already_signed = self
-            .signatures
-            .iter()
-            .any(|signature| signature.public_key == *public_key);
-        if !already_signed {
-            self.signatures.push(Signature::new(
+        if self.signatures.contains_key(public_key) {
+            Err("Already signed by these keys.".to_string())
```
Comment (humb1t) [inline iroha/src/sumeragi.rs:151]: Do we need error in such a case? Maybe we can just leave as is or update signature?

File: iroha/src/sumeragi.rs [hunk 11]
```diff
@@ -132,7 +203,9 @@ pub struct Sumeragi {
     /// PendingBlock in discussion this round
     voting_block: Arc<RwLock<Option<VotingBlock>>>,
     blocks_sender: Arc<RwLock<ValidBlockSender>>,
+    transaction_sender: TransactionSender,
     world_state_view: Arc<RwLock<WorldStateView>>,
+    pending_forwarded_tx_hashes: Arc<RwLock<BTreeSet<Hash>>>,
```
Comment (humb1t) [inline iroha/src/sumeragi.rs:208]: This field should be commented or renamed

File: iroha/src/sumeragi.rs (hunk 9 above)
Comment (eadventurous) [reply iroha/src/sumeragi.rs:64]: I guess we should do it in a separate refactoring Pull Request

File: iroha/src/sumeragi.rs (hunk 10 above)
Comment (eadventurous) [reply iroha/src/sumeragi.rs:151]: We don't need an explicit error, but we do need a result. We should know if we added a new signature or not, to decide whether to forward the message.

File: iroha/src/sumeragi.rs (hunk 9 above)
Comment (humb1t) [reply iroha/src/sumeragi.rs:64]: ok

File: iroha/src/config.rs [hunk 12]
```diff
@@ -21,11 +21,17 @@ const TRUSTED_PEERS: &str = "IROHA_TRUSTED_PEERS";
...
 const IROHA_PUBLIC_KEY: &str = "IROHA_PUBLIC_KEY";
 const IROHA_PRIVATE_KEY: &str = "IROHA_PRIVATE_KEY";
+const COMMIT_TIME_MS: &str = "COMMIT_TIME_MS";
+const TX_RECEIPT_TIME_MS: &str = "TX_RECEIPT_TIME_MS";
 const DEFAULT_TORII_URL: &str = "127.0.0.1:1337";
 const DEFAULT_BLOCK_TIME_MS: u64 = 1000;
 const DEFAULT_KURA_INIT_MODE: Mode = Mode::Strict;
 const DEFAULT_KURA_BLOCK_STORE_PATH: &str = "./blocks";
 const DEFAULT_MAX_FAULTY_PEERS: usize = 0;
+/// Amount of time Peer waits for CommitMessage from the proxy tail.
```
Comment (humb1t) [inline iroha/src/config.rs:31]: We don't have `CommitMessage` entity

**Pull request merged**